File: axl/compiler.py
import os
import re
from dataclasses import dataclass, fields, is_dataclass, replace
from pathlib import Path

from .compact import CompactParseError, is_compact_source, split_compact_frames
from .ir import Function, FunctionCall, Program
from .parser import ParseError, parse


class CompileError(ValueError):
    """Raised when source files cannot be assembled into one AXL program."""


_IMPORT = re.compile(r'^\s*import\s+([A-Za-z_][A-Za-z0-9_]*)\s+from\s+"([^"\n]+)"\s*$')
_RESERVED = {
    "agent",
    "call",
    "else",
    "emit",
    "end",
    "false",
    "fn",
    "forget",
    "if",
    "import",
    "let",
    "memory",
    "meta",
    "recall",
    "return",
    "run",
    "true",
    "uses",
    "while",
    "workflow",
}
_MAX_IMPORT_DEPTH = 256
_MAX_MODULES = 1_024
_MAX_TOTAL_SOURCE_BYTES = 4 * 1024 * 1024


@dataclass
class _CompileContext:
    root: Path
    modules: int = 0
    source_bytes: int = 0


def compile_file(path: str | Path, *, module_root: str | Path | None = None) -> Program:
    entry = Path(path).resolve()
    root = Path(module_root).resolve() if module_root is not None else entry.parent
    return _compile_file(entry, (), _CompileContext(root))


def _compile_file(
    path: Path, stack: tuple[Path, ...], context: _CompileContext
) -> Program:
    if len(stack) > _MAX_IMPORT_DEPTH:
        raise CompileError(f"import depth exceeds {_MAX_IMPORT_DEPTH}")
    if not path.is_relative_to(context.root):
        raise CompileError(f"module '{path.name}' is outside module root")
    if path.suffix != ".axl":
        raise CompileError("modules must use the .axl extension")
    if path in stack:
        chain = " -> ".join(item.name for item in (*stack, path))
        raise CompileError(f"cyclic module import: {chain}")
    context.modules += 1
    if context.modules > _MAX_MODULES:
        raise CompileError(f"module count exceeds {_MAX_MODULES}")
    remaining = _MAX_TOTAL_SOURCE_BYTES - context.source_bytes
    try:
        with path.open("rb") as stream:
            size = os.fstat(stream.fileno()).st_size
            if size > remaining:
                raise CompileError(
                    f"aggregate module source exceeds {_MAX_TOTAL_SOURCE_BYTES} bytes"
                )
            data = stream.read(remaining + 1)
    except CompileError:
        raise
    except OSError as error:
        raise CompileError(f"cannot read module '{path}': {error}") from error
    if len(data) > remaining:
        raise CompileError(
            f"aggregate module source exceeds {_MAX_TOTAL_SOURCE_BYTES} bytes"
        )
    try:
        source = data.decode("utf-8")
    except UnicodeError as error:
        raise CompileError(f"cannot read module '{path}': {error}") from error
    context.source_bytes += len(data)

    imports: list[tuple[str, Path]] = []
    aliases: set[str] = set()
    try:
        if is_compact_source(source):
            local_source = _compact_imports(
                source, path, imports, aliases, context.root
            )
        else:
            local_source = _legacy_imports(source, path, imports, aliases, context.root)
        local_program = parse(local_source)
    except (CompactParseError, ParseError) as error:
        raise CompileError(f"{path}: {error}") from error

    imported_instructions = []
    for alias, imported_path in imports:
        imported = _compile_file(imported_path, (*stack, path), context)
        if not all(
            isinstance(instruction, Function) for instruction in imported.instructions
        ):
            raise CompileError(
                f"module '{imported_path}' may only export function declarations"
            )
        imported_instructions.extend(_namespace(imported, alias).instructions)

    return Program((*imported_instructions, *local_program.instructions))


def _compact_imports(
    source: str,
    path: Path,
    imports: list[tuple[str, Path]],
    aliases: set[str],
    module_root: Path,
) -> str:
    frames = split_compact_frames(source)
    local_frames = [frames[0]]
    depth = 0
    for index, frame in enumerate(frames[1:], 1):
        opcode = frame.split("|", 1)[0].strip()
        if opcode == "1":
            if depth:
                raise CompileError(
                    f"{path}:frame {index}: compact import must be top-level"
                )
            fields = [field.strip() for field in frame.split("|")]
            if len(fields) != 3:
                raise CompileError(f"{path}:frame {index}: invalid compact import")
            alias, relative_path = fields[1:]
            _add_import(
                alias, relative_path, path, index, imports, aliases, module_root
            )
            continue
        local_frames.append(frame)
        if opcode in {"30", "32", "40", "50", "51", "61"}:
            depth += 1
        elif opcode == "99" and depth:
            depth -= 1
    return ";".join(local_frames)


def _legacy_imports(
    source: str,
    path: Path,
    imports: list[tuple[str, Path]],
    aliases: set[str],
    module_root: Path,
) -> str:
    local_lines: list[str] = []
    depth = 0
    for number, line in enumerate(source.splitlines(), 1):
        stripped = line.strip()
        if not stripped.startswith("import "):
            local_lines.append(line)
            if stripped == "end" and depth:
                depth -= 1
            elif stripped.startswith(("if ", "while ", "agent ", "fn ", "workflow ")):
                depth += 1
            continue
        if depth:
            raise CompileError(f"{path}:{number}: import must be top-level")
        match = _IMPORT.fullmatch(line)
        if match is None:
            raise CompileError(f"{path}:{number}: invalid import declaration")
        alias, relative_path = match.groups()
        _add_import(alias, relative_path, path, number, imports, aliases, module_root)
        local_lines.append("")
    return "\n".join(local_lines)


def _add_import(
    alias: str,
    relative_path: str,
    path: Path,
    position: int,
    imports: list[tuple[str, Path]],
    aliases: set[str],
    module_root: Path,
) -> None:
    if not re.fullmatch(r"[A-Za-z_][A-Za-z0-9_]*", alias) or alias in _RESERVED:
        raise CompileError(f"{path}:{position}: invalid import alias '{alias}'")
    if alias in aliases:
        raise CompileError(f"duplicate import alias '{alias}'")
    if not relative_path or "|" in relative_path or ";" in relative_path:
        raise CompileError(f"{path}:{position}: invalid import path")
    candidate = Path(relative_path)
    if candidate.is_absolute():
        raise CompileError(f"{path}:{position}: import path must be relative")
    if ".." in candidate.parts:
        raise CompileError(f"{path}:{position}: import path escapes module root")
    imported_path = (path.parent / candidate).resolve()
    if not imported_path.is_relative_to(module_root):
        raise CompileError(f"{path}:{position}: import path escapes module root")
    if imported_path.suffix != ".axl":
        raise CompileError(f"{path}:{position}: imported module must use .axl")
    aliases.add(alias)
    imports.append((alias, imported_path))


def _namespace(program: Program, alias: str) -> Program:
    names = {
        instruction.name: f"{alias}.{instruction.name}"
        for instruction in program.instructions
        if isinstance(instruction, Function)
    }

    def transform(value):
        if isinstance(value, FunctionCall) and value.name in names:
            value = replace(value, name=names[value.name])
        if isinstance(value, Function) and value.name in names:
            value = replace(value, name=names[value.name])
        if is_dataclass(value):
            updates = {
                field.name: transform(getattr(value, field.name))
                for field in fields(value)
            }
            return replace(value, **updates)
        if isinstance(value, tuple):
            return tuple(transform(item) for item in value)
        return value

    return transform(program)

File: runtime/axl-core-rs/src/compiler.rs
use std::collections::{HashMap, HashSet};
use std::path::{Path, PathBuf};

use crate::compact::{parse_compact, split_compact_frames, CompactParseError};
use crate::ir::*;
use crate::validation;

#[derive(Debug)]
pub struct CompileError(pub String);

impl std::fmt::Display for CompileError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl std::error::Error for CompileError {}

const MAX_IMPORT_DEPTH: usize = 32;
const MAX_MODULES: usize = 256;
const MAX_SOURCE_BYTES: usize = 4 * 1024 * 1024;

struct CompileContext {
    root: PathBuf,
    modules: usize,
    source_bytes: usize,
    imported: HashSet<PathBuf>,
}

/// Compile an AXL file with module support
pub fn compile_file(path: &Path) -> Result<Program, CompileError> {
    let root = path.parent().unwrap_or(Path::new(".")).to_path_buf();
    let mut ctx = CompileContext {
        root, modules: 0, source_bytes: 0,
        imported: HashSet::new(),
    };
    compile_file_inner(path, &mut ctx)
}

fn compile_file_inner(path: &Path, ctx: &mut CompileContext) -> Result<Program, CompileError> {
    if ctx.modules > MAX_MODULES {
        return Err(CompileError(format!("module count exceeds {MAX_MODULES}")));
    }
    if ctx.imported.contains(path) {
        return Err(CompileError(format!("cyclic import: {}", path.display())));
    }
    ctx.imported.insert(path.to_path_buf());
    ctx.modules += 1;

    let source = std::fs::read_to_string(path)
        .map_err(|e| CompileError(format!("cannot read '{}': {e}", path.display())))?;

    if source.len() > MAX_SOURCE_BYTES {
        return Err(CompileError("source exceeds 4MB".into()));
    }
    ctx.source_bytes += source.len();

    // Parse imports
    let (imports, local_source) = extract_imports(&source, path, &ctx.root)?;

    // Parse local program
    let mut program = parse_compact(&local_source)
        .map_err(|e| CompileError(format!("{}: {e}", path.display())))?;

    // Process imports
    for (alias, import_path) in &imports {
        let imported = compile_file_inner(import_path, ctx)?;

        // Namespace imported functions
        for instruction in imported.instructions {
            match instruction {
                Instruction::Function(func) => {
                    let namespaced = Function {
                        name: format!("{alias}.{}", func.name),
                        ..func
                    };
                    program.instructions.insert(0, Instruction::Function(namespaced));
                }
                _ => {}
            }
        }
    }

    Ok(program)
}

fn extract_imports(
    source: &str,
    path: &Path,
    root: &Path,
) -> Result<(Vec<(String, PathBuf)>, String), CompileError> {
    let mut imports = Vec::new();
    let mut aliases = HashSet::new();
    let mut local_frames = Vec::new();

    let frames = split_compact_frames(source)
        .map_err(|e| CompileError(format!("{}: {e}", path.display())))?;

    if frames.is_empty() || (frames[0] != "2" && frames[0] != "3") {
        return Err(CompileError("compact source requires version header '2' or '3'".into()));
    }
    local_frames.push(frames[0].clone());

    let mut depth = 0;
    for (i, frame) in frames[1..].iter().enumerate() {
        let fields: Vec<&str> = frame.split('|').collect();
        let opcode = fields.first().copied().unwrap_or("");

        if opcode == "1" && fields.len() == 3 {
            if depth > 0 {
                return Err(CompileError(format!("{}: frame {}: import must be top-level", path.display(), i + 1)));
            }
            let alias = fields[1];
            let relative_path = fields[2];

            if aliases.contains(alias) {
                return Err(CompileError(format!("{}: frame {}: duplicate import alias '{alias}'", path.display(), i + 1)));
            }

            let import_path = path.parent().unwrap_or(Path::new(".")).join(relative_path);
            let resolved = import_path.canonicalize()
                .map_err(|e| CompileError(format!("{}: frame {}: cannot resolve '{relative_path}': {e}", path.display(), i + 1)))?;

            if !resolved.starts_with(root) {
                return Err(CompileError(format!("{}: frame {}: import path escapes module root", path.display(), i + 1)));
            }

            aliases.insert(alias.to_string());
            imports.push((alias.to_string(), resolved));
        } else {
            local_frames.push(frame.clone());
            if matches!(opcode, "30" | "32" | "40" | "50" | "51") {
                depth += 1;
            } else if opcode == "99" && depth > 0 {
                depth -= 1;
            }
        }
    }

    Ok((imports, local_frames.join(";")))
}

/// List all functions in a program
pub fn list_functions(program: &Program) -> Vec<String> {
    program.instructions.iter().filter_map(|i| {
        if let Instruction::Function(f) = i {
            Some(f.name.clone())
        } else {
            None
        }
    }).collect()
}

/// List all agents in a program
pub fn list_agents(program: &Program) -> Vec<String> {
    program.instructions.iter().filter_map(|i| {
        if let Instruction::Agent(a) = i {
            Some(a.name.clone())
        } else {
            None
        }
    }).collect()
}

/// List all workflows in a program
pub fn list_workflows(program: &Program) -> Vec<String> {
    program.instructions.iter().filter_map(|i| {
        if let Instruction::Workflow(w) = i {
            Some(w.name.clone())
        } else {
            None
        }
    }).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn compile_simple_file() {
        let dir = std::env::temp_dir().join("axl_compile_test");
        std::fs::create_dir_all(&dir).unwrap();
        let file = dir.join("test.axl");
        std::fs::write(&file, "2;12|\"hello\"").unwrap();

        let program = compile_file(&file).unwrap();
        assert_eq!(program.instructions.len(), 1);

        std::fs::remove_dir_all(&dir).unwrap();
    }

    #[test]
    fn list_functions_works() {
        let program = Program { instructions: vec![
            Instruction::Function(Function {
                name: "test".into(), parameters: vec![], return_type: "int".into(), body: vec![],
            }),
            Instruction::Emit(Expression::Literal(Value::Int(1))),
        ] };
        let funcs = list_functions(&program);
        assert_eq!(funcs.len(), 1);
        assert_eq!(funcs[0], "test");
    }

    #[test]
    fn list_agents_works() {
        let program = Program { instructions: vec![
            Instruction::Agent(Agent {
                name: "research".into(), tools: vec![], body: vec![],
                goal: None, tool_defs: vec![], memory_defs: vec![], handlers: vec![],
            }),
        ] };
        let agents = list_agents(&program);
        assert_eq!(agents.len(), 1);
        assert_eq!(agents[0], "research");
    }
}
